contrib/bindings/go/pathrs/pathrs.go

```diff
@@ -0,0 +1,294 @@
+}
+
+// Free frees underling caught resources
+func (h *Handle) Free() {
+	if h == nil {
+		return
+	}
+
+	free(ptrTypeHandle, h.handle)
+}
+
+// Reopen upgrade the handle to a file representation
+func (h *Handle) Reopen(flags int) (*os.File, error) {
+	handlePtr := (*C.pathrs_handle_t)(h.handle)
+	fd := fd(C.pathrs_reopen(handlePtr, C.int(flags)))
+
+	if err := handleErr(ptrTypeHandle, h.handle); err != nil {
+		return nil, err
+	}
+
+	file := os.NewFile(uintptr(fd), h.uniqFileName())
+	return file, nil
+}
+
+func (h *Handle) uniqFileName() string {
+	name := fmt.Sprintf("pathrs:%d", h.fileCounter)
+	h.fileCounter++
+	return name
+}
+
+// Error representation of rust error
+// particularly useful to not frighten to lost controll of pointer which can be rewritten.
+type Error struct {
+	description string
+	errno       uint64
+	backtrace   []backtraceLine
+}
+
+type backtraceLine struct {
+	ip       uintptr
+	sAddress uintptr
+	sName    string
+	sFile    string
+	sLineno  uint32
+}
+
+func (err *Error) Error() string {
+	return err.description
+}
+
+func (e *Error) Unwrap() error {
+	if e.errno != 0 {
+		return syscall.Errno(e.errno)
+	}
+
+	return nil
+}
+
+// Backtrace flush backtrace of underlying error to string.
+//
+// Its not passed to realization of Error interface on purpose since
+// the main error should remain clear and simple
+func (err *Error) Backtrace() string {
+	buf := strings.Builder{}
+	buf.Reset()
+
+	for _, line := range err.backtrace {
+		if line.sName != "" {
+			buf.WriteString(fmt.Sprintf("'%s'@", line.sName))
+		}
+		buf.WriteString(fmt.Sprintf("<0x%x>+0x%x\n", line.sAddress, line.ip-line.sAddress))
+		if line.sFile != "" {
+			buf.WriteString(fmt.Sprintf("  in file '%s':%d\n", line.sFile, line.sLineno))
+		}
+	}
+
+	return buf.String()
+}
+
+func newError(e cErrorPtr) *Error {
+	realError := (*errorType)(e)
+
+	err := &Error{}
+	err.errno = uint64(realError.savedErrno)
+	if realError.description != nil {
+		err.description = C.GoString(realError.description)
+	}
+
+	if realError.backtrace != nil {
+		length := uintptr(realError.backtrace.length)
+		size := unsafe.Sizeof(backtraceEntry{})
+		for i := uintptr(0); i < length; i++ {
+			head := (*backtraceEntry)(unsafe.Pointer(uintptr(unsafe.Pointer(realError.backtrace.head)) + size*i))
+			line := backtraceLine{}
+			line.sAddress = head.symbolAddress
+			line.ip = head.ip
+			if head.symbolName != nil {
+				line.sName = C.GoString(head.symbolName)
+			}
+			if head.symbolFile != nil {
+				line.sFile = C.GoString(head.symbolFile)
+				line.sLineno = uint32(head.symbolLineno)
+			}
+
+			err.backtrace = append(err.backtrace, line)
+		}
+	}
+
+	return err
+}
+
+func handleErr(t pointerType, ptr unsafe.Pointer) (err error) {
+	ptrType := C.pathrs_type_t(t)
+	errPtr := cErrorPtr(C.pathrs_error(ptrType, ptr))
+	if errPtr != nil {
+		err = newError(errPtr)
+	}
+	free(ptrTypeError, errPtr)
+
+	return
+}
+
+func free(ptrType pointerType, ptr unsafe.Pointer) {
```

Comment:
Why not just make this take a `C.pathrs_type_t`? I find it harder to follow as-is, and there's no real benefit to having the Go versions of these constants (since they're only used inside the binding).